// vectra/lib/LocalIndex.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_assert_1 = __importDefault(require("node:assert"));
const sinon_1 = __importDefault(require("sinon"));
const LocalIndex_1 = require("./LocalIndex");
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
describe('LocalIndex', () => {
    const testIndexDir = path_1.default.join(__dirname, 'test_index');
    const basicIndexItems = [
        { id: '1', vector: [1, 2, 3] },
        { id: '2', vector: [2, 3, 4] },
        { id: '3', vector: [3, 4, 5] }
    ];
    beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
        yield promises_1.default.rm(testIndexDir, { recursive: true, force: true });
    }));
    afterEach(() => __awaiter(void 0, void 0, void 0, function* () {
        yield promises_1.default.rm(testIndexDir, { recursive: true, force: true });
        sinon_1.default.restore();
    }));
    it('should create a new index', () => __awaiter(void 0, void 0, void 0, function* () {
        const index = new LocalIndex_1.LocalIndex(testIndexDir);
        yield index.createIndex();
        const created = yield index.isIndexCreated();
        node_assert_1.default.equal(created, true);
        node_assert_1.default.equal(index.folderPath, testIndexDir);
    }));
    it('blocks concurrent operations when lock is held', () => __awaiter(void 0, void 0, void 0, function* () {
        const index = new LocalIndex_1.LocalIndex(testIndexDir);
        yield index.createIndex();
        yield index.beginUpdate(); // grab lock for a big update!
        yield node_assert_1.default.rejects(() => __awaiter(void 0, void 0, void 0, function* () {
            yield index.beginUpdate(); // try to grab lock again. should fail!
        }), new Error('Update already in progress'));
    }));
    describe('createIndex', () => {
        it('checks for existing index on creation', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex(); // create first index.json
            // create without deleteIfExists. Will reject
            yield node_assert_1.default.rejects(() => __awaiter(void 0, void 0, void 0, function* () {
                yield index.createIndex();
            }), new Error('Index already exists'));
            // create with deleteIfExists. Should remove old data
            yield index.insertItem({ id: '1', vector: [1, 2, 3] });
            const lengthBefore = (yield index.listItems()).length;
            node_assert_1.default.equal(lengthBefore, 1);
            yield index.createIndex({ deleteIfExists: true, version: 2, metadata_config: {} });
            const lengthAfter = (yield index.listItems()).length;
            node_assert_1.default.equal(lengthAfter, 0);
        }));
        it('delete index if file creation fails', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            sinon_1.default.stub(promises_1.default, 'writeFile').rejects(new Error('fs error'));
            yield node_assert_1.default.rejects(() => __awaiter(void 0, void 0, void 0, function* () {
                yield index.createIndex();
            }), new Error('Error creating index'));
            yield node_assert_1.default.rejects(() => __awaiter(void 0, void 0, void 0, function* () {
                yield index.listItems();
            }));
        }));
    });
    describe('deleteItem', () => {
        it('does nothing when id not found', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            yield index.beginUpdate();
            yield index.insertItem(basicIndexItems[0]);
            yield index.insertItem(basicIndexItems[1]);
            yield index.insertItem(basicIndexItems[2]);
            yield index.endUpdate();
            yield node_assert_1.default.doesNotReject(() => __awaiter(void 0, void 0, void 0, function* () {
                yield index.deleteItem('dne');
            }));
            node_assert_1.default.equal((yield index.listItems()).length, 3);
        }));
        it('leaves existing empty index when last el deleted', () => __awaiter(void 0, void 0, void 0, function* () {
            var _a;
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            yield index.insertItem(basicIndexItems[0]);
            yield index.deleteItem((_a = basicIndexItems[0].id) !== null && _a !== void 0 ? _a : '');
            node_assert_1.default.equal(yield index.isIndexCreated(), true);
            node_assert_1.default.equal((yield index.listItems()).length, 0);
        }));
        it('removes elements from any position', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            yield index.batchInsertItems([
                { id: '1', vector: [] },
                { id: '2', vector: [] },
                { id: '3', vector: [] },
                { id: '4', vector: [] },
                { id: '5', vector: [] },
            ]);
            yield index.beginUpdate();
            yield index.deleteItem('1');
            yield index.deleteItem('3');
            yield index.deleteItem('5');
            yield index.endUpdate();
            node_assert_1.default.deepStrictEqual(yield index.listItems(), [{ id: '2', vector: [], metadata: {}, norm: 0 }, { id: '4', vector: [], metadata: {}, norm: 0 }]);
        }));
    });
    describe('endUpdate', () => {
        it('throws an error if no update has begun', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield node_assert_1.default.rejects(() => __awaiter(void 0, void 0, void 0, function* () {
                yield index.endUpdate();
            }), new Error('No update in progress'));
        }));
        it('throws an error if the index could not be saved', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir, 'index.json');
            yield index.createIndex();
            yield index.beginUpdate();
            sinon_1.default.stub(promises_1.default, 'writeFile').rejects(new Error('fs error'));
            yield node_assert_1.default.rejects(() => __awaiter(void 0, void 0, void 0, function* () {
                yield index.endUpdate();
            }), new Error('Error saving index: Error: fs error'));
        }));
    });
    describe('getIndexStats', () => {
        it('reports empty index correctly', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            node_assert_1.default.deepStrictEqual(yield index.getIndexStats(), {
                version: 1,
                metadata_config: {},
                items: 0
            });
        }));
        it('correctly reports non-empty index stats', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex({ version: 1, metadata_config: { indexed: [] } });
            yield index.batchInsertItems(basicIndexItems);
            node_assert_1.default.deepStrictEqual(yield index.getIndexStats(), {
                version: 1,
                metadata_config: { indexed: [] },
                items: 3
            });
        }));
    });
    describe('getItem', () => {
        it('returns undefined when item not found', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            node_assert_1.default.equal(yield index.getItem('1'), undefined);
        }));
        it('returns requested item', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            yield index.batchInsertItems(basicIndexItems);
            const item2 = yield index.getItem('2');
            node_assert_1.default.equal(item2 === null || item2 === void 0 ? void 0 : item2.id, basicIndexItems[1].id);
            node_assert_1.default.equal(item2 === null || item2 === void 0 ? void 0 : item2.vector, basicIndexItems[1].vector);
            node_assert_1.default.equal((yield index.listItems()).length, 3);
        }));
    });
    describe('batchInsertItems', () => {
        it('should insert provided items', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            const newItems = yield index.batchInsertItems(basicIndexItems);
            node_assert_1.default.equal(newItems.length, 3);
            const retrievedItems = yield index.listItems();
            node_assert_1.default.equal(retrievedItems.length, 3);
        }));
        it('on id collision - cancel batch insert & bubble up error', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            yield index.insertItem({ id: '2', vector: [9, 9, 9] });
            // ensures insert error is bubbled up to batchIndexItems caller
            yield node_assert_1.default.rejects(() => __awaiter(void 0, void 0, void 0, function* () {
                yield index.batchInsertItems(basicIndexItems);
            }), {
                name: 'Error',
                message: 'Item with id 2 already exists'
            });
            // ensures no partial update is applied
            const storedItems = yield index.listItems();
            node_assert_1.default.equal(storedItems.length, 1);
        }));
    });
    describe('listItemsByMetadata', () => {
        it('returns items matching metadata filter', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            yield index.batchInsertItems([
                { id: '1', vector: [], metadata: { category: 'food' } },
                { id: '2', vector: [], metadata: { category: 'food' } },
                { id: '3', vector: [], metadata: { category: 'electronics' } },
                { id: '4', vector: [], metadata: { category: 'drink' } },
                { id: '5', vector: [], metadata: { category: 'food' } },
            ]);
            const foodItems = yield index.listItemsByMetadata({ category: { '$eq': 'food' } });
            node_assert_1.default.deepStrictEqual(foodItems.map((item) => item.id), ["1", "2", "5"]);
            const drinkItems = yield index.listItemsByMetadata({ category: { '$eq': 'drink' } });
            node_assert_1.default.deepStrictEqual(drinkItems.map((item) => item.id), ["4"]);
            const clothingItems = yield index.listItemsByMetadata({ category: { '$eq': 'clothes' } });
            node_assert_1.default.deepStrictEqual(clothingItems, []);
        }));
        it('returns nothing when no items in index', () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            const items = yield index.listItemsByMetadata({});
            node_assert_1.default.deepStrictEqual(items, []);
        }));
    });
    describe("queryItems", () => {
        it("returns empty array on empty index search", () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            const result = yield index.queryItems([1, 2, 3], "", 10);
            node_assert_1.default.deepStrictEqual(result, []);
        }));
        it("returns bad match when no better match exists", () => __awaiter(void 0, void 0, void 0, function* () {
            var _a, _b;
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            yield index.insertItem({ id: "1", vector: [0.9, 0, 0, 0, 0] });
            const result = yield index.queryItems([0, 0, 0, 0, 0.1], "", 1);
            node_assert_1.default.equal((_a = result[0]) === null || _a === void 0 ? void 0 : _a.score, 0);
            node_assert_1.default.equal((_b = result[0]) === null || _b === void 0 ? void 0 : _b.item.id, "1");
        }));
        it("returns all vectors when fewer than topK exist", () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            yield index.batchInsertItems(basicIndexItems);
            const result = yield index.queryItems([0, 0, 1], "", 10);
            node_assert_1.default.equal(result.length, 3);
            node_assert_1.default.deepStrictEqual(result.map(({ item }) => item.id), basicIndexItems.map((item) => item.id));
        }));
        it("filters by metadata when filter provided", () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex();
            yield index.batchInsertItems([
                { id: "1", vector: [1, 0, 0], metadata: { category: "food" } },
                { id: "2", vector: [0, 0, 1], metadata: { category: "drink" } },
            ]);
            const bestGeneralMatch = yield index.queryItems([1, 0, 0], "", 1);
            const bestDrinkMatch = yield index.queryItems([1, 0, 0], "", 1, {
                category: { $eq: "drink" },
            });
            node_assert_1.default.equal(bestGeneralMatch[0].item.id, "1");
            node_assert_1.default.equal(bestDrinkMatch[0].item.id, "2");
        }));
        it("reads item metadata file when provided", () => __awaiter(void 0, void 0, void 0, function* () {
            const index = new LocalIndex_1.LocalIndex(testIndexDir);
            yield index.createIndex({ version: 1, metadata_config: { indexed: ['category'] } });
            yield index.batchInsertItems([
                { id: "1", vector: [1, 0, 0] },
                { id: "2", vector: [0, 0, 1], metadata: { category: 'drink' } },
            ]);
            sinon_1.default
                .stub(promises_1.default, "readFile")
                .resolves(JSON.stringify({ category: "drink" }));
            const bestDrinkMatch = yield index.queryItems([1, 0, 0], "", 2, { category: { '$eq': 'drink' } });
            node_assert_1.default.notEqual(bestDrinkMatch[0].item.metadataFile, undefined);
            node_assert_1.default.equal(bestDrinkMatch[0].item.id, "2");
        }));
    });
});
